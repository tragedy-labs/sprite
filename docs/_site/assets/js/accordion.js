/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
