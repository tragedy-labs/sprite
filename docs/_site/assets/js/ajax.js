// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
