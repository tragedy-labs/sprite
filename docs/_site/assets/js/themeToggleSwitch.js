function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
