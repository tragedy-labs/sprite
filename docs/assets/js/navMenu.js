document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
