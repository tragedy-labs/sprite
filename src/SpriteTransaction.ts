import { SpriteDatabase } from './SpriteDatabase.js';

export class SpriteTransaction {
  private _id: string;
  private _committed: boolean = false;
  private database: SpriteDatabase;
  constructor(database: SpriteDatabase, transactionId: string) {
    this.database = database;
    this._id = transactionId;
  }
  /** The trasaction ID */
  get id() {
    return this._id;
  }
  /** Whether or not the transaction has been committed */
  get committed() {
    return this._committed;
  }
  commit = async () => {
    // TODO: This old committed logic is flawed and perhaps replacing
    // with a status would be a better approach.
    this._committed = await this.database.commitTransaction(this.id);
    return this.committed;
  };
  rollback = async () => this.database.rollbackTransaction(this.id);
}
