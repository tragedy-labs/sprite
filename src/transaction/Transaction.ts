import { DatabaseSession } from '../session/DatabaseSession.js';
import { ArcadeSupportedQueryLanguages } from '../database/Database.js';
import { Rest } from '../rest/Rest.js';
import { Routes } from '../database/routes.js';
import { SpriteTransaction } from './SpriteTransaction.js';
import { HeaderKeys } from '@/rest/SpriteHeaders.js';

/**
 * isolationLevel is the isolation level for the current transaction,
 * either `READ_COMMITTED` or `REPEATABLE_READ`. For details on what
 * isolation level dictates about the transaction, see the [ArcadeDB
 * documentation](https://docs.arcadedb.com/#HTTP-Begin)
 * @default READ_COMMITTED
 */
export type ArcadeTransactionIsolationLevel =
  | 'READ_COMMITTED'
  | 'REPEATABLE_READ';

/**
 * Static class for performing operations on or in a transaction.
 */
export class Transaction {
  /**
   * Static method to begin a transaction.
   * @param session The {@link DatabaseSession `DatabaseSession`} to begin the transaction on.
   * @param isolationLevel The isolation level of the transaction.
   * @returns A {@link SpriteTransaction `SpriteTransaction`} instance for the target session.
   */
  public static begin = async (
    session: DatabaseSession,
    isolationLevel?: ArcadeTransactionIsolationLevel
  ): Promise<SpriteTransaction> => {
    try {
      // 'READ_COMMITTED' is default in ARCADEDB,
      // so we don't bother sending that
      const response = await Rest.post(
        Routes.BEGIN,
        isolationLevel === 'REPEATABLE_READ'
          ? JSON.stringify({ isolationLevel })
          : null,
        session
      );

      if (response.status !== 204) {
        throw new Error(
          `Server returned an unexpected response. Status: ${response.status} / ${response.statusText}.`
        );
      }

      const transactionId = response.headers?.get(HeaderKeys.ArcadeSessionId);

      // because the headers could be null
      // TODO: this isn't a good way to check for the transactionId
      if (!transactionId || typeof transactionId !== 'string') {
        throw new Error('Invalid transaction key received from server.');
      } else {
        return new SpriteTransaction(session, transactionId);
      }
    } catch (error) {
      throw new Error(
        `Unable to begin transaction in database "${session.databaseName}".`,
        { cause: error }
      );
    }
  };
  /**
   * Static method to commit a transaction.
   * @param session The {@link DatabaseSession `DatabaseSession`} to commit the transaction on.
   * @param transaction The id of the transaction to commit.
   * @returns `true` if the transaction was committed.
   */
  public static commit = async (
    session: DatabaseSession,
    transaction: SpriteTransaction
  ) => {
    try {
      const result = await Rest.post(Routes.COMMIT, null, session, transaction);
      if (result.status === 204) {
        return true;
      } else {
        throw new Error(
          `Unexpected response from the server when attemping to commit transaction ${transaction.id}`
        );
      }
    } catch (error) {
      throw new Error(`Unable to commit transaction ${transaction.id}`, {
        cause: error
      });
    }
  };
  /**
   * Static method to manage a transaction within the scope.
   * @param session The {@link DatabaseSession `DatabaseSession`} to execute the transaction on.
   * @param callback The callback to execute within transaction context.
   * @param isolationLevel The {@link ArcadeTransactionIsolationLevel isolationLevel} of the transaction.
   * @returns A touple containing result of the transaction, and a boolean indicating if the transaction was committed or rolled-back.
   */
  public static manage = async <T = void>(
    session: DatabaseSession,
    callback: (trx: SpriteTransaction) => Promise<T>,
    isolationLevel?: ArcadeTransactionIsolationLevel
  ): Promise<[boolean, T]> => {
    const trx = await this.begin(session, isolationLevel);
    try {
      const result = await callback(trx);
      if (trx.rolledBack) {
        return [false, result];
      } else {
        await trx.commit();
        return [true, result];
      }
    } catch (error) {
      // Attempt to rollback the transaction
      try {
        await trx.rollback();
        throw new Error(`Transaction ${trx.id} failed and was rolled back.`, {
          cause: error
        });
      } catch (rollbackError) {
        throw new Error(
          `Transaction ${trx.id} failed and rollback also failed.`,
          { cause: rollbackError }
        );
      }
    }
  };
  /**
   * Static method to execute CRUD as part of a transaction.
   * @param language The {@link ArcadeSupportedQueryLanguages `ArcadeSupportedQueryLanguages`} to use for the command.
   * @param command The CRUD command to execute.
   * @param params The parameters to pass to the command.
   * @returns The result of the CRUD operation.
   * @throws Error if the CRUD operation fails.
   */
  public static crud = async <T>(
    session: DatabaseSession,
    transaction: SpriteTransaction,
    language: ArcadeSupportedQueryLanguages,
    command: string,
    params?: Record<string, boolean | string | number>
  ): Promise<T> => {
    try {
      return await Rest.postJson<T>(
        Routes.COMMAND,
        { language, command, params },
        session,
        transaction
      );
    } catch (error) {
      throw new Error(`Could not perform CRUD operation ${command}`, {
        cause: error
      });
    }
  };
}
