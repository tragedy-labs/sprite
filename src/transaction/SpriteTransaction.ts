import { DatabaseSession } from '../session/DatabaseSession.js';
import { ArcadeSupportedQueryLanguages } from '../database/Database.js';
import { Transaction } from './Transaction.js';

/**
 * A transaction in Sprite, contains the transaction id, and methods to
 * commit or rollback the transaction.
 * @param session the `DatabaseSession` the transaction is taking place of
 * @param transactionId the id of the transaction
 */
export class SpriteTransaction {
  private _session: DatabaseSession;
  private _id: string;
  /** State */
  private _rolledBack = false;
  private _committed = false;
  constructor(session: DatabaseSession, transactionId: string) {
    this._session = session;
    this._id = transactionId;
  }
  /** The trasaction ID */
  get id() {
    return this._id;
  }
  /** Whether or not the transaction has been rolledback */
  get rolledBack() {
    return this._rolledBack;
  }
  /** Whether or not the transaction has been committed */
  get committed() {
    return this._committed;
  }
  /**
   * Commit the transaction.
   * @returns `true` if the transaction was commited.
   */
  commit = async () => {
    this._committed = await Transaction.commit(this._session, this);
    return this.committed;
  };
  /**
   * Rollback the transaction.
   * @returns `true` if the transaction was commited.
   */
  rollback = async () => {
    this._rolledBack = await Transaction.rollback(this._session, this);
    return this.rolledBack;
  };
  /**
   * Perform a CRUD operation in the transaction.
   * @param language the query language to use
   * @param command the command to execute
   * @param params the (optional) parameters to pass to the command
   * @returns the result of the CRUD operation
   */
  crud = async <T>(
    language: ArcadeSupportedQueryLanguages,
    command: string,
    parameters?: Record<string, boolean | string | number>
  ): Promise<T> =>
    Transaction.crud<T>(this._session, this, language, command, parameters);
}
