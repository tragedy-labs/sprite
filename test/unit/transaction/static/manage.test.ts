// Lib
import { SpriteTransaction } from '@/transaction/SpriteTransaction.js';
import { Transaction } from '@/transaction/Transaction.js';

// Testing
import { TestDatabaseSession as SESSION, variables } from '@test/variables.js';

describe('SpriteTransaction.manage()', () => {
  beforeEach(() => {
    jest
      .spyOn(Transaction, 'commit')
      .mockImplementationOnce(async (): Promise<boolean> => {
        return true;
      });
  });
  it(`correctly passes isolationLevel argument to SpriteDatabase.newTransaction`, async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    await Transaction.manage(SESSION, async () => {}, 'REPEATABLE_READ');

    expect(Transaction.begin).toHaveBeenCalledWith(SESSION, 'REPEATABLE_READ');
  });

  it(`correctly passes a new SpriteTransaction to the callback`, async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    const [, transaction] = await Transaction.manage<SpriteTransaction>(
      SESSION,
      async (trx: SpriteTransaction) => {
        return trx;
      }
    );

    expect(transaction).toBeInstanceOf(SpriteTransaction);
  });

  it(`correctly commits the transaction before returning`, async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    const [commited] = await Transaction.manage(SESSION, async () => {});
    expect(commited).toBe(true);
  });

  it(`it executes the callback once`, async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    let count = 0;
    await Transaction.manage(SESSION, async () => {
      ++count;
    });

    expect(count).toBe(1);
  });

  // TODO: Maybe the boolean value should be the first value in the touple
  it('if the transaction is rolled back it returns a touple containing the transactions return value, and a boolean false value', async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    jest
      .spyOn(Transaction, 'rollback')
      .mockImplementationOnce(async (): Promise<boolean> => {
        return true;
      });

    const [bool, trx] = await Transaction.manage(
      SESSION,
      async (trx: SpriteTransaction) => {
        await trx.rollback();
        return trx;
      }
    );
    expect(bool).toBe(false);
    expect(trx).toBeInstanceOf(SpriteTransaction);
    expect(trx.rolledBack).toBe(true);
  });

  it('if an error occurs within the callback it rolls back the transaction', async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    jest
      .spyOn(Transaction, 'rollback')
      .mockImplementationOnce(async (): Promise<boolean> => {
        return true;
      });

    await expect(
      Transaction.manage(SESSION, async () => {
        throw new Error('Simulated error');
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws an error if an error occurs within the callback', async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });
    const transactionPromise = Transaction.manage(SESSION, () => {
      throw new Error('Simulated error');
    });
    await expect(transactionPromise).rejects.toMatchSnapshot();
  });

  it('rollsback the transaction if an error occurs within the error handling', async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    jest
      .spyOn(Transaction, 'rollback')
      .mockImplementationOnce(async (): Promise<boolean> => {
        return true;
      });

    let transaction: SpriteTransaction | undefined;

    const transactionPromise = Transaction.manage(
      SESSION,
      async (trx: SpriteTransaction) => {
        transaction = trx;
        throw new Error('Simulated error');
      }
    );

    await expect(transactionPromise).rejects.toThrowErrorMatchingSnapshot();
    expect(transaction?.rolledBack).toBe(true);
  });

  it('throws an error if the transaction fails to rollback in the error handling', async () => {
    jest
      .spyOn(Transaction, 'begin')
      .mockImplementationOnce(async (): Promise<SpriteTransaction> => {
        return new SpriteTransaction(SESSION, variables.sessionId);
      });

    jest
      .spyOn(Transaction, 'rollback')
      .mockImplementationOnce(async (): Promise<boolean> => {
        throw new Error('Simulated error');
      });

    await expect(
      Transaction.manage(SESSION, async () => {
        throw new Error('Simulated error');
      })
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
