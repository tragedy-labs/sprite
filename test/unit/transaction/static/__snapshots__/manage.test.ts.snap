// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteTransaction.manage() if an error occurs within the callback it rolls back the transaction 1`] = `"Transaction AS-0000000-0000-0000-0000-00000000000 failed and rollback also failed."`;

exports[`SpriteTransaction.manage() rollsback the transaction if an error occurs within the error handling 1`] = `"Transaction AS-0000000-0000-0000-0000-00000000000 failed and rollback also failed."`;

exports[`SpriteTransaction.manage() throws an error if an error occurs within the callback 1`] = `[Error: Transaction AS-0000000-0000-0000-0000-00000000000 failed and rollback also failed.]`;

exports[`SpriteTransaction.manage() throws an error if the transaction fails to rollback in the error handling 1`] = `"Transaction AS-0000000-0000-0000-0000-00000000000 failed and rollback also failed."`;
